src/mod_muc_room.erl

```diff
@@ -1317,12 +1317,34 @@ set_affiliation(JID, Affiliation, StateData, Reason) ->
     Room = StateData#state.room,
     Host = StateData#state.host,
     Mod = gen_mod:db_mod(ServerHost, mod_muc),
-    case Mod:set_affiliation(ServerHost, Room, Host, JID, Affiliation, Reason) of
-	ok ->
-	    StateData;
-	{error, _} ->
-	    set_affiliation_fallback(JID, Affiliation, StateData, Reason)
-    end.
+    LUser = JID#jid.luser,
+    NewAffiliation = case {Affiliation, Reason} of
+    {outcast, <<"muted">>} ->
+        muted;
+    _ ->
+        Affiliation
+    end,
+    case NewAffiliation of
+    none ->
+        case ets_cache:lookup(user_affiliation_cache, LUser) of
```

Comment:
I believe that this should be abstracted into get_affiliation and just call get_affiliation.

Reply:
I think previous caches are 'hidden in this way

Reply:
Good point